src/main/scala/edu/ie3/util/osm/model/ContainerCache.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * © 2022. TU Dortmund University,
+ * Institute of Energy Systems, Energy Efficiency and Energy Economics,
+ * Research group Distribution grid planning and operation
+*/
+package edu.ie3.util.osm.model
+
+import com.typesafe.scalalogging.LazyLogging
+import edu.ie3.util.exceptions.OsmException
+import edu.ie3.util.osm.OsmUtils
+import edu.ie3.util.osm.model.OsmContainer.ParOsmContainer.SimpleParOsmContainer
+import edu.ie3.util.osm.model.OsmContainer.SeqOsmContainer.SimpleSeqOsmContainer
+import edu.ie3.util.osm.model.OsmEntity.ComposedEntity.{ExtendedWay, SimpleWay}
+import edu.ie3.util.osm.model.OsmEntity.ComposedEntity.Relation.{
+  ExtendedRelation,
+  RelationMemberType,
+  SimpleRelation
+}
+import edu.ie3.util.osm.model.OsmEntity.ComposedEntity.Relation.RelationMember.ExtendedRelationMember
+import edu.ie3.util.osm.model.OsmEntity.ComposedEntity.Way.{
+  ExtendedWay,
+  SimpleWay
+}
+import edu.ie3.util.osm.model.OsmEntity.{
+  ExtendedOsmEntity,
+  Node,
+  SimpleOsmEntity
+}
+
+import scala.collection.immutable.{MapOps, OldHashMap}
+import scala.collection.{GenMap, mutable}
+import scala.collection.parallel.immutable.{ParMap, ParMapLike}
+import scala.util.{Failure, Success, Try}
+
+trait ContainerCache extends LazyLogging {
+
+  protected def _simpleWay: Long => Option[SimpleWay]
+
+  protected def _simpleRelation: Long => Option[SimpleRelation]
+
+  protected def _node: Long => Option[Node]
+
+  protected val _wayCache: mutable.Map[Long, ExtendedWay] =
+    scala.collection.mutable.Map.empty
+
+  protected val _relCache: mutable.Map[Long, ExtendedRelation] =
+    scala.collection.mutable.Map.empty
+
+  /** WARNING: METHOD IS NOT THREAD SAFE AND MAY CAUSE OVERHEAD IN GENERATING
+    * THE SAME WAY MULTIPLE TIMES
+    *
+    * @param id
+    * @return
+    */
+  def extendedWay(id: Long): Option[ExtendedWay] =
+    _wayCache.get(id) match {
+      case cachedWay @ Some(_) =>
+        cachedWay
+      case None =>
+        _simpleWay(id)
+          .flatMap(_.asExtended(_node) match {
+            case Failure(exception) =>
+              logger.error(
+                s"Cannot convert SimpleWay '$id' to ExtendedWay. " +
+                  s"Exception: $exception."
+              )
```

Comment:
```suggestion
              logger.error(
                s"Cannot convert SimpleWay '$id' to ExtendedWay. " , exception
              )
```

AFAIK, if you provide the exception as a second parameter, you get a better stacktrace support in log file